google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/facet/flexible/FlexibleFacetEditor.java

```diff
@@ -0,0 +1,347 @@
+    );
+
+    yaml.getTextField().getDocument().addDocumentListener(new DocumentAdapter() {
+      @Override
+      protected void textChanged(DocumentEvent event) {
+        toggleDockerfileSection();
+        validateConfiguration();
+      }
+    });
+
+    dockerfile.getTextField().getDocument().addDocumentListener(new DocumentAdapter() {
+      @Override
+      protected void textChanged(DocumentEvent event) {
+        toggleDockerfileSection();
+        validateConfiguration();
+      }
+    });
+
+    dockerfile.addBrowseFolderListener(
+        GctBundle.message("appengine.flex.config.browse.dockerfile"),
+        null /* description */,
+        project,
+        FileChooserDescriptorFactory.createSingleFileDescriptor()
+    );
+
+    genYamlButton.addActionListener(new GenerateConfigActionListener(project, "app.yaml",
+        appEngineHelper::defaultAppYaml, yaml, this::validateConfiguration));
+
+    genDockerfileButton.addActionListener(new GenerateConfigActionListener(project, "Dockerfile",
+        () -> appEngineHelper.defaultDockerfile(AppEngineFlexibleDeploymentArtifactType.WAR),
+        dockerfile, this::validateConfiguration
+    ));
+
+    yaml.setText(deploymentConfiguration.getYamlPath());
+    dockerfile.setText(deploymentConfiguration.getDockerFilePath());
+
+    errorIcon.setIcon(Ide.Error);
+    errorIcon.setVisible(false);
+    errorMessage.setVisible(false);
+
+    validateConfiguration();
+    toggleDockerfileSection();
+  }
+
+  @NotNull
+  @Override
+  public JComponent createComponent() {
+    return mainPanel;
+  }
+
+  @Override
+  public boolean isModified() {
+    return !yaml.getText().equals(deploymentConfiguration.getYamlPath())
+        || !dockerfile.getText().equals(deploymentConfiguration.getDockerFilePath());
+  }
+
+  @Override
+  public void reset() {
+    yaml.setText(deploymentConfiguration.getYamlPath());
+    dockerfile.setText(deploymentConfiguration.getDockerFilePath());
+
+    toggleDockerfileSection();
+  }
+
+  @Override
+  public void apply() throws ConfigurationException {
+    validateConfiguration();
+    if (!isValidConfigurationFile(yaml.getText())) {
+      throw new ConfigurationException(
+          GctBundle.getString("appengine.deployment.error.staging.yaml"));
+    }
+
+    if (isRuntimeCustom() && !isValidConfigurationFile(dockerfile.getText())) {
+      throw new ConfigurationException(
+          GctBundle.getString("appengine.deployment.error.staging.dockerfile"));
+    }
+
+    deploymentConfiguration.setYamlPath(yaml.getText());
+    deploymentConfiguration.setDockerFilePath(dockerfile.getText());
+
+    toggleDockerfileSection();
+  }
+
+  @Override
+  public void disposeUIResources() {
+    // Do nothing.
+  }
+
+  @Nls
+  @Override
+  public String getDisplayName() {
+    return GctBundle.getString("appengine.flexible.facet.name");
+  }
+
+  private boolean isRuntimeCustom() {
+    return APP_ENGINE_PROJECT_SERVICE.getFlexibleRuntimeFromAppYaml(yaml.getText())
```

Comment:
You mean the project service isn't the best place for it?
I was looking for the best "utils" class and the project service one sounded the best, since it already does similar appengine-web.xml parsing for Standard.